ci: workflows: pin python dependencies

nashif opened the pull request:
Pin python dependencies to hashes and cleanup/unify python setup steps in
various workflows.

We now have one dependency file containing all requirements for github
actions that is managed centrally with hashes. No direct pip installs
are needed in workflow files and everything shall go via the
requirements file.

Pinning to specific version and hashes helps with preventing supply
chain attacks.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
ci: workflows: pin python dependencies

Pin python dependencies to hashes and cleanup/unify python setup steps in
various workflows.

We now have one dependency file containing all requirements for github
actions that is managed centrally with hashes. No direct pip installs
are needed in workflow files and everything shall go via the
requirements file.

Pinning to specific version and hashes helps with preventing supply
chain attacks.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

aescolar commented on kernel/init.c:
```diff
@@ -1,3 +1,4 @@
+
```
This slipped in?

aescolar commented on .github/workflows/backport_issue_check.yml:
```diff
@@ -28,16 +28,27 @@ jobs:
+        run: |
+          pip install -r scripts/requirements-actions.txt --require-hashes
+
+      - name: install-packages
```
dup

pdgendt commented on scripts/requirements-actions.in:
```diff
@@ -0,0 +1,40 @@
+pygithub
```
Sort the entries and add
```
# zephyr-keep-sorted-start
...
# zephyr-keep-sorted-stop
```

kartben replied:
I was actually going to suggest something different and try to make an effort at grouping/documenting some of these so that we at least have some sense of what workflows / class of workflows requires what. Like group all the linters, all the tools for postprocessing test reports (junit, elasticsearch, xlsx, ...), ...

pdgendt replied:
I'd put the same comment if it was split into groups, not sure how easy we can categorize them.

pdgendt commented:
> hmm, not sure about this one:
> 
> ```
>  Traceback (most recent call last):
>   File "/opt/hostedtoolcache/Python/3.12.9/x64/bin/gitlint", line 5, in <module>
>     from gitlint.cli import cli
>   File "/opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages/gitlint/cli.py", line 3, in <module>
>     from gitlint.config import LintConfig, LintConfigError
>   File "/opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages/gitlint/config.py", line 2, in <module>
>     import ConfigParser
> ModuleNotFoundError: No module named 'ConfigParser'
> Compliance error, check for error messages in the "Run Compliance Tests" st
> ```
> 
> I have not seen this during testing :(

Can we update the `gitlint` package? Also for the `requirements.in` file, we could specify `gitlint-core` instead.

pdgendt commented on scripts/requirements-actions.txt:
```diff
@@ -0,0 +1,1320 @@
+# This file is autogenerated by pip-compile with Python 3.12
+# by the following command:
+#
+#    pip-compile --allow-unsafe --generate-hashes --output-file=requirements-actions.txt requirements-actions.in
```
Maybe add `--upgrade`?

kartben commented:
I think you will need to update dependabot configuration as well https://github.com/dependabot/dependabot-core/pull/10040#issuecomment-2696978430

nashif commented:
> I think you will need to update dependabot configuration as well [dependabot/dependabot-core#10040 (comment)](https://github.com/dependabot/dependabot-core/pull/10040#issuecomment-2696978430)

hmm, is this reallyneeded? we do have a requirements.txt file in this case. Anyways, one step at a time, I do not think we need to change depandabot workflow for this to go in, no?

kartben commented:
> > I think you will need to update dependabot configuration as well [dependabot/dependabot-core#10040 (comment)](https://github.com/dependabot/dependabot-core/pull/10040#issuecomment-2696978430)
> 
> hmm, is this reallyneeded? we do have a requirements.txt file in this case. Anyways, one step at a time, I do not think we need to change depandabot workflow for this to go in, no?

Well I would argue we do. This effectively freezes all dependencies (duh!) so should any one of them be affected by a CVE 1/ we won't know about it and 2/, and more importantly, we will never get the fix that may come in a later version and will just be hard-stuck with the vulnerable version, where the current approach would at least have had the benefit to always give us the latest version of the package (for better or worse, too).

So what I'm saying is that basically we will need dependabot eventually if we are to pin dependencies ; and while it can be added in a PR coming shortly after this one, we do need to at least test it works and be confident it will be happy with managing uv compiled requirements before introducing this change in the first place.

In any case, CI seems to be failing so maybe that'll be the opportunity to add this to the next update :)

nashif committed:
tests: twister: disable test_inline_logs temporarily

This test fails on older python versions (3.10) and only on CI.
Disabling it while we investigate. The test itself verifies inline logs
options, so the functionality test is not impacted.

Tracked in #87769

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

cfriedt commented:
Should help with caching / false positives as well.

pdgendt commented:
Why was the "Checkout source code" step skipped in the failed "Publish Unit Tests Results"? 🤔

nashif commented:
> Why was the "Checkout source code" step skipped in the failed "Publish Unit Tests Results"? 🤔

this is fixed now

pdgendt commented:
The commit to temporarily disable the failing test also touches the `.github/workflows/twister.yaml` file to checkout the pull request head. Is this intentional?

nashif commented:
> The commit to temporarily disable the failing test also touches the `.github/workflows/twister.yaml` file to checkout the pull request head. Is this intentional?

no, will fix,  thanks.

nashif committed:
ci: overhaul and pin dependencies on twister_tests_blackbox

Pin dependencies on the workflow and move it from using docker to the
zephyr setup action.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
> > The commit to temporarily disable the failing test also touches the `.github/workflows/twister.yaml` file to checkout the pull request head. Is this intentional?
> 
> no, will fix, thanks.

ok, fixed and removed the workaround.

kartben commented on .github/workflows/bsim-tests.yaml:
```diff
@@ -178,7 +178,6 @@ jobs:
 
       - name: Merge Test Results
         run: |
-          pip install junitparser junit2html
```
Looks like you just dropped them as I can't see any step where you install the requirements?

nashif replied:
those are already in the docker image, no need to install them

kartben commented on .github/workflows/clang.yaml:
```diff
@@ -135,13 +135,30 @@ jobs:
+          cache: pip
+          cache-dependency-path: scripts/requirements-actions.txt
+
+      - name: install-packages
```
If you have a chance to do a `s/install-packages/Install Python packages/g` so that steps have consistent naming that would be great (non blocking obviously)

kartben commented on .github/workflows/scripts_tests.yml:
```diff
@@ -37,37 +37,16 @@ jobs:
           ref: ${{ github.event.pull_request.head.sha }}
           fetch-depth: 0
 
-      - name: Rebase
-        continue-on-error: true
-        env:
-          BASE_REF: ${{ github.base_ref }}
-          PR_HEAD: ${{ github.event.pull_request.head.sha }}
-        run: |
-          git config --global user.email "actions@zephyrproject.org"
-          git config --global user.name "Github Actions"
-          rm -fr ".git/rebase-apply"
-          rm -fr ".git/rebase-merge"
-          git rebase origin/${BASE_REF}
-          git clean -f -d
-          git log --graph --oneline HEAD...${PR_HEAD}
-
```
Unrelated change?

teburd commented:
Lets do it with the entire environment using nix :-D Hash all the things!

pdgendt commented:
Missed it last time I reviewed, but commit messages shouldn't contain Github `#number` references 🫣
